gateway/gateway.go

```diff
@@ -161,6 +179,22 @@ func NewGateway(
 	return g, nil
 }
 
+// look at the resource.Config to determine which model is being used.
+// Alternatively, we could add an enum to the gateway struct and let the NewGateway functions specify which model is being used.
+func getNativeConfig(conf resource.Config) (*Config, error) {
+	// check if we are the generic config
+	cfg, err := resource.NativeConfig[*Config](conf)
+	if err == nil {
+		return cfg, nil
+	}
+	// check if we are the waveshare hat
+	waveshareHatCfg, err := resource.NativeConfig[*ConfigSX1302WaveshareHAT](conf)
+	if err != nil {
+		return nil, fmt.Errorf("the config %v does not match a supported config type", conf)
+	}
+	return waveshareHatCfg.getGatewayConfig(), nil
+}
+
```

Comment:
in hardware testing & tests, this function worked. I couldn't trigger a situation where the `resource.Config` was getting written to the wrong struct. 

Can switch this to using enums with a switch if people prefer that.

Reply:
We can just have a common provate creator get passed in the config pins as board objects, and an spi bus name? those are the three potential differences in code?

Reply:
I do not understand what you mean. 

I needed this function because I wanted the waveshare_hat model to have a different json Config, so the current config validation functions wouldn't have to change.

gps does not require this function because they are an [always rebuild](https://github.com/viam-modules/gps/blob/7ca4edc68afd04213b5637ba3829a78c60201dec/rtk/gpsrtk.go#L40) model, but since we have `Reconfigure` we need to handle parsing different configs.

Reply:
Ah yes Reconfigure means my plan doesn't work. I was thinking of having a common private function for both gateways that take in a reset pin and power enable pin as `board.GPIOPin` objects and a spi bus string, since that's the common functionality between the two.

But I did not consider reconfiguring, which takes in the two configuration objects. 

I like your enum idea a lot more in this case - it will be neat to implement a case structure on the Config type.

Thanks for the successful mind reading.

Reply:
Kk updated. Realized I could use the model/model name in the resource.Config for this so I like where it landed